FitnessTrackerApi core development

Georgiy-111 opened the pull request:
Удалил IsDeleted из методов в WorkoutService.cs и из модели Workout.cs , выдавало ошибку 500 в swagger. конфликт между логикой "мягкого удаления" (soft delete) и настройками Entity Framework.

Georgiy-111 committed:
Удалил IsDeleted из методов в WorkoutService.cs и из модели Workout.cs , выдавало ошибку 500 в swagger. конфликт между логикой "мягкого удаления" (soft delete) и настройками Entity Framework.

Georgiy-111 committed:
Сменил базу данных с MySQL на PostgreSQL, так же изменил методы в WorkoutService.cs ,не работал поиск/удаление и обновление по id в swagger, теперь никаких ошибок нет и таблицы создаются/удаляются/редактируются и просматриваются корректно.

Georgiy-111 committed:
Создал интерфейс `IWorkoutRepository` с методами для работы с тренировками.
 Реализовал класс `WorkoutRepository`, который использует `ApplicationDbContext` для взаимодействия с базой данных.
 Зарегистрировал `WorkoutRepository` в DI контейнере в `Program.cs`.
 Так же добавил недостающие xml комментарии.

Georgiy-111 committed:
Добавил несколько проверок в WorkoutService.cs и привел их к одному общему виду.

o5breadfan commented on Repositories/IWorkoutRepository.cs:
```diff
@@ -0,0 +1,35 @@
+    
+    /// <summary>
+    /// Удаляет тренировку по ID.
+    /// </summary>
```
```suggestion
    /// </summary>
    <param ...
```
Добавь атрибут <param /> на все методы где есть параметры. По примеру как в workoutrepository у конструктора

o5breadfan commented on Repositories/WorkoutRepository.cs:
```diff
@@ -0,0 +1,101 @@
+    /// Получает все тренировки.
+    /// </summary>
+    /// <param name="cancellationToken">Токен для отмены операции.</param>
+    /// <returns>Список всех тренировок.</returns>
```
```suggestion
    /// <returns>Список всех тренировок.</returns>
```
А вот эти summary убери. Пусть только в интерфейсе будет описание

o5breadfan commented on Service/IWorkoutService.cs:
```diff
@@ -8,10 +8,30 @@ namespace FitnessTrackerApi.Service;
+    
+    /// <summary>
+    /// Получает все тренировки.
+    /// </summary>
```
```suggestion
    /// </summary>
    /// <param ...>
```
Точно также добавить параметры как и выше писал

o5breadfan commented on Service/IWorkoutService.cs:
```diff
@@ -8,10 +8,30 @@ namespace FitnessTrackerApi.Service;
+    /// Удаляет тренировку по ID.
+    /// </summary>
     Task<bool> DeleteAsync(int id, CancellationToken cancellationToken);
     
```
```suggestion
    
```
лишний пробел

o5breadfan commented on Service/WorkoutService.cs:
```diff
@@ -44,11 +39,17 @@ public async Task<IEnumerable<Workout>> GetAllAsync(CancellationToken cancellati
         {
             if (id <= 0)
```
```suggestion
            if (id <= 0)
```
Проверка ни о чем не говорит. Кроме как о том, что злоумышленник знает что айдишник должен быть больше 0)) В принципе если мы знаем, что таких идентификаторов у нас быть не может и никогда не будет, то проще возвращать null. Ну то есть тренировка с таким айдишником не найдена.
Пусть будет лучше return null; в этом месте

o5breadfan commented on appsettings.json:
```diff
@@ -1,6 +1,6 @@
 {
   "ConnectionStrings": {
-    "DefaultConnection": "server=127.0.0.1;port=3306;database=fitness_tracker_db;user=root;password=Ger31960956mexico"
+    "DefaultConnection": "Host=localhost;Port=5432;Database=fitness_tracker_db;Username=postgres;Password=Ger31960956"
```
```suggestion
    "DefaultConnection": "Host=localhost;Port=5432;Database=fitness_tracker_db;Username=postgres;Password=Ger31960956"
```
У тебя есть appsettings.Development.json вынеси туда раздел ConnectionStrings в котором у тебя локальные подключения. В файле appsettings обычно содержаться конфигурации с прода. Еще бывают конфигурации:
appsettings.Test.json - конфигурации с тестового окружения (там разворачивается ветка develop и тестировщики могут месить говно всякое там, менять данные, создавать пользователей типо "Тестов тест Тестович" и так далее
appsettings.Stage.json - конфигурации с стейджа (препрод). Это окружение в котором максимально приближено к проду (возможно даже данные с прода копировались на этот стенд) Там проверяют перед релизами. Так как там чистые данные как на проде итд.
appsettings.Production.json - конфигурация с прода. Тут уже все серьезно)) 
appsettings.json - это по сути главный файл конфигурации. Все остальные зависимые. Например если в appsettings.json есть раздел ConnectionStrings, а в appsettings.Development.json такого раздела нет, то при запуске приложения с конфигурацией appsettings.Development.json, приложение автоматически подтянет этот раздел из родительского appsettings.json.

o5breadfan replied:
Как понять с какой конфигурацией запускается приложение? Если откроешь файл launchSettings.json в папке Properties, то увидишь там раздел profiles (http, https, IIS Express) это по сути профили запуска приложения. Во всех этих профилях есть раздел environmentVariables, в котором и указывается с какой конфигурацией будет запускаться приложение. 
Если Development, то конфигурация будет браться из appsettings.Development.json
Если Stage, то конфигурация будет браться из appsettings.Stage.json
Если Production, то из appsettings.Production.json
Если huemoe, то из appsettings.huemoe.json.

Сделай 4 конфигурации (Development, Test, Stage, Production) и сделай 4 профиля одноименных с разными переменными окружения (environmentVariables)

Georgiy-111 committed:
— Добавил атрибуты на все методы где есть параметры;
- Убрал лишние дублирующие описания, в основном оставил summary в интерфейсах;
— Обновил в WorkoutService.cs GetByIdAsync (null, если не найден);
— Вынес ConnectionStrings в appsettings.Development.json;
— добавил конфигурации и launch профили для окружений(Development, Test, Stage, Production);
- В Test, Stage, Production оставил просто заглушки, не знаю нужно ли было дублировать ConnectionStrings или делать как то по другому, поэтому пока не трогал;

Georgiy-111 committed:
-summary только в интерфейсах;

Georgiy-111 committed:
- Вынес регистрацию сервисов, репозиториев и подключения БД в методы-расширения;
- Добавлены методы-расширения AddCoreServices, AddRepositories и AddDatabase для IServiceCollection;
- Сделал метод-расширение GetDatabaseConnection для IConfiguration;

Georgiy-111 committed:
- Удалил DbContextExtensions, перенес от туда метод AddDatabase в ServiceCollectionExtensions.cs;

Georgiy-111 committed:
- Реализован CRUD для сущности WorkoutType: сущность, 3 DTO, репозиторий, сервис, контроллер;
- Добавлены интерфейсы и реализации: IWorkoutTypeRepository, WorkoutTypeRepository, IWorkoutTypeService, WorkoutTypeService;
- Настроены зависимости через методы-расширения AddRepositories и AddCoreServices;
- Добавлены миграции: AddWorkoutTypeTable и обновлена БД;

Georgiy-111 committed:
Migrations

o5breadfan commented on Controllers/Admin/WorkoutTypeController.cs:
```diff
@@ -0,0 +1,62 @@
+
+namespace FitnessTrackerApi.Controllers.Admin;
+
+[Route("api/admin/[controller]")]
```
> [Route("api/admin/[controller]")]

оформление уехало

o5breadfan commented on Controllers/Admin/WorkoutTypeController.cs:
```diff
@@ -0,0 +1,62 @@
+        }
+        
+        [HttpGet]
+        public async Task<IActionResult> GetAll(CancellationToken cancellationToken)
```
Добавь summary на каждый метод, чтобы в сваггере было описание что это для чего и какой результат он возвращает

o5breadfan commented on Controllers/Admin/WorkoutTypeController.cs:
```diff
@@ -0,0 +1,62 @@
+        }
+        
+        [HttpPost]
+        public async Task<IActionResult> Create(WorkoutTypeCreateDto dto, CancellationToken cancellationToken)
```
> WorkoutTypeCreateDto dto

У тебя скорее приходит ViewModel с фронта (UI), и потом ты эту ViewModel маппишь в dto, т.к. dto может содержать и другие поля, которые не заполняет фронт. Лучше назвать WorkoutTypeCreateRequestModel и тут использовать маппинг из WorkoutTypeCreateRequestModel в WorkoutTypeCreateRequest.
Для этого используй AutoMapper (в хороших местах все плюются от него, но все равно встречается почти везде)

o5breadfan commented on DTOs/WorkoutType/WorkoutTypeCreateDto.cs:
```diff
@@ -0,0 +1,10 @@
+public class WorkoutTypeCreateDto
+{
+    [Required(ErrorMessage = "Название обязательно!")]
+    [MaxLength(100, ErrorMessage = "Максимальная длина 100 символов!")]
```
> [Required(ErrorMessage = "Название обязательно!")]
    [MaxLength(100, ErrorMessage = "Максимальная длина 100 символов!")]

Вместо этих атрибутов, везде используй библиотеку FluentValidation. Посмотри как с ней работать

o5breadfan commented on Program.cs:
```diff
@@ -1,8 +1,11 @@
 
 var builder = WebApplication.CreateBuilder(args);
 
+builder.Services.AddScoped<IApplicationDbContext, ApplicationDbContext>();
```
не тут должно быть

o5breadfan commented on Models/WorkoutType.cs:
```diff
@@ -0,0 +1,13 @@
+    [MaxLength(100, ErrorMessage = "Максимальная длина 100 символов!")]
+    public string Name { get; set; } = string.Empty;
+    
+   // public ICollection<WorkoutSchedule> WorkoutSchedules { get; set; } = new List<WorkoutSchedule>();
```
неиспользуемый код

o5breadfan commented on Repositories/Interfaces/IWorkoutTypeRepository.cs:
```diff
@@ -0,0 +1,16 @@
+
+public interface IWorkoutTypeRepository
+{
+    Task<IEnumerable<WorkoutType>> GetAllAsync();
```
нет описания методов

o5breadfan replied:
И нет CancellationToken'ов

o5breadfan commented on Service/Interfaces/IWorkoutTypeService.cs:
```diff
@@ -0,0 +1,16 @@
+
+public interface IWorkoutTypeService
+{
+    Task<IEnumerable<WorkoutTypeDto>> GetAllAsync(CancellationToken cancellationToken);
```
Описание методов

o5breadfan commented on Service/WorkoutTypeService.cs:
```diff
@@ -0,0 +1,92 @@
+
+public class WorkoutTypeService : IWorkoutTypeService
+{
+    private readonly IApplicationDbContext _context;
```
Почему доступ к БД идет из сервиса, а не репозитория

Georgiy-111 committed:
-Исправлено оформление в WorkoutTypeController.cs;
-Добавлен summary на каждый метод, что бы в swagger было описание.
-Убрал атрибуты по типу [MaxLength] и [Required], вместо их использовал FluentValidation;
-Вынес IApplicationDbContext с Program.cs в ServiceCollectionExtensions.cs;
-Добавил описания методов и CancellationToken в IWorkoutTypeRepository.cs;
-Добавил описания методов в IWorkoutTypeService.cs;
-В WorkoutTypeService.cs доступ к БД был через сервис, исправил, теперь через репозиторий;
(Осталось добавить некоторые xml комментарии и разобраться с AutoMapper )

o5breadfan commented on Controllers/Admin/WorkoutTypeController.cs:
```diff
@@ -5,58 +5,91 @@
+    /// </summary>
+    /// <param name="id">Идентификатор типа тренировки.</param>
+    /// <param name="cancellationToken">Токен отмены операции.</param>
+    /// <returns>Тип тренировки или 404, если не найден.</returns>
```
в returns пишется то что возвращается при HTTP Status'е = 200. Все остальное (404 и прочие ошибки) указываются дополнительным атрибутом над методом (как [HttpGet] - это атрибут) [ProducesResponseType ...]

o5breadfan commented on Extensions/ConfigurationExtensions.cs:
```diff
@@ -2,7 +2,7 @@ namespace FitnessTrackerApi.Extensions
     public static class ConfigurationExtensions
     {
-        public static string GetDatabaseConnection(this IConfiguration configuration)
+        public static string? GetDatabaseConnection(this IConfiguration configuration)
```
Не должно быть Nullable типа. Всегда должна возвращаться строго строка. Если он тебе говорит что там может быть null, то значит это надо обработать, нельзя позволять вытаскивать вместо строки подключения какую-то хрень. То есть если строка = null то выбрасывать исключение, что строка подключения не найдена.

o5breadfan commented on Data/Interfaces/IApplicationDbContext.cs:
```diff
@@ -0,0 +1,12 @@
+using FitnessTrackerApi.Models;
```
Зачем тебе этот интерфейс ? В классе уже все есть, как и метод SaveChangesAsync. Удали этот интерфейс и используй класс. Ты когда регистрировал базу данных в ServiceProvider путем .AddDbContext(....) то если посмотреть реализацию этого метода, то внутри он делает насколько помню .AddScoped<ApplicationDbContext>(....) Поэтому ты мягко говоря дважды его зарегистрировал, как просто класс (при подключении БД), так и в паре с интерфейсом от которого якобы он зависит. Поэтому в данный момент это излишне

o5breadfan commented on Extensions/ServiceCollectionExtensions.cs:
```diff
@@ -0,0 +1,36 @@
+        {
+            services.AddDbContext<ApplicationDbContext>(options =>
+                options.UseNpgsql(configuration.GetDatabaseConnection()));
+            services.AddScoped<IApplicationDbContext, ApplicationDbContext>();
```
Ну и соответственно уберешь это когда удалишь интерфейс.

>             services.AddScoped<IApplicationDbContext, ApplicationDbContext>();

o5breadfan commented on Validators/WorkoutTypeCreateDtoValidator.cs:
```diff
@@ -0,0 +1,14 @@
+{
+    public WorkoutTypeCreateDtoValidator()
+    {
+        RuleFor(x => x.Name)
```
Мало правил, добавь еще. Плюс ты проверил что они работают ? пытался через сваггер запрос отправить на создание тренировки с пустым именем ?

o5breadfan replied:
Ну и напиши нормальный текст ошибки)

o5breadfan commented on Repositories/WorkoutRepository.cs:
```diff
@@ -0,0 +1,80 @@
+            .ToListAsync(cancellationToken);
+    }
+    
+    /// <param name="id">Идентификатор тренировки.</param>
```
Когда говорил убрать summary это значит вообще весь XML комментарии вместе с param returns и прочее. У тебя это все описано в интерфейсе. В реализации (в классе) это не надо

Georgiy-111 committed:
-Добавил ProducesResponseType в WorkoutTypeController.cs и WorkoutController.cs;
-Поправил ConfigurationExtensions.cs;
-Удалил IApplicationDbContext.cs и его AddScoped в ServiceCollectionExtensions.cs;
-Добавил правила в WorkoutTypeCreateDtoValidator.cs;
-Убрал лишние описания(комментарии) в реализации и оставил их в интерфейсах;
-Настроил AutoMapper: профиль, маппинги и регистрация;

Georgiy-111 committed:
-Добавил явный маппинг полей через ForMember в AutoMapper;
